File: CfxGenerator/ApiTypes/CefEnumType.vb
'' Copyright (c) 2014-2015 Wolfgang Borgsmüller
'' All rights reserved.
'' 
'' Redistribution and use in source and binary forms, with or without 
'' modification, are permitted provided that the following conditions 
'' are met:
'' 
'' 1. Redistributions of source code must retain the above copyright 
''    notice, this list of conditions and the following disclaimer.
'' 
'' 2. Redistributions in binary form must reproduce the above copyright 
''    notice, this list of conditions and the following disclaimer in the 
''    documentation and/or other materials provided with the distribution.
'' 
'' 3. Neither the name of the copyright holder nor the names of its 
''    contributors may be used to endorse or promote products derived 
''    from this software without specific prior written permission.
'' 
'' THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
'' "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
'' LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
'' FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
'' COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
'' INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
'' BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS 
'' OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND 
'' ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
'' TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE 
'' USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Public Class CefEnumType
    Inherits ApiType

    Public Class EnumMember
        Public ReadOnly Name As String
        Public ReadOnly Value As String
        Public ReadOnly Comments As CommentData

        Sub New(ByVal name As String, ByVal value As String, Comments As CommentData)
            Me.Name = name
            If value.Length > 0 Then Me.Value = value
            Me.Comments = Comments
        End Sub
        Public Overrides Function ToString() As String
            If Value IsNot Nothing Then
                Return Name & " = " & Value
            Else
                Return Name
            End If
        End Function
    End Class

    Private ReadOnly members As EnumMember()
    Private ReadOnly comments As CommentData

    Sub New(data As Parser.EnumData)
        MyBase.New(data.Name)
        members = New EnumMember(data.Members.Count - 1) {}
        For i = 0 To data.Members.Count - 1
            members(i) = New EnumMember(data.Members(i).Name, data.Members(i).Value, data.Members(i).Comments)
        Next
        comments = data.Comments
    End Sub

    Protected Sub New(name As String)
        MyBase.New(name)
    End Sub

    Public Overrides ReadOnly Property OriginalSymbol As String
        Get
            Return Name & "_t"
        End Get
    End Property

    Public Overrides ReadOnly Property ProxySymbol As String
        Get
            Return "int"
        End Get
    End Property

    Public Overrides ReadOnly Property RemoteSymbol As String
        Get
            Return PublicSymbol
        End Get
    End Property

    Public ReadOnly Property CfxName As String
        Get
            Return "cfx_" & Name.Substring(4)
        End Get
    End Property

    Public Overrides ReadOnly Property PInvokeSymbol As String
        Get
            Return CSharp.ApplyStyle(CfxName)
        End Get
    End Property

    Public Overrides ReadOnly Property RemoteUnwrapExpression(var As String) As String
        Get
            Return "(int)" & var
        End Get
    End Property

    Public Overrides ReadOnly Property RemoteWrapExpression(var As String) As String
        Get
            Return String.Format("({0}){1}", RemoteSymbol, var)
        End Get
    End Property

    Public Overrides ReadOnly Property ProxyWrapExpression(var As String) As String
        Get
            Return "(int)" & var
        End Get
    End Property

    Public Overrides ReadOnly Property ProxyUnwrapExpression(var As String) As String
        Get
            Return String.Format("({0}){1}", PublicSymbol, var)
        End Get
    End Property


    Private Shared additionalFlags As String() = {"CfxV8PropertyAttribute"}

    Public Sub EmitEnum(b As CodeBuilder)

        Dim enumName = CSharp.ApplyStyle(CfxName)

        b.AppendSummaryAndRemarks(comments)

        If Name.Contains("_flags") OrElse additionalFlags.Contains(enumName) Then
            b.AppendLine("[Flags()]")
        End If

        Dim varPrefix = members(0).Name
        For i = 1 To members.Length - 1
            If varPrefix.Length > members(i).Name.Length Then
                varPrefix = varPrefix.Substring(0, members(i).Name.Length)
            End If
            For c = 0 To varPrefix.Length - 1
                If varPrefix(c) <> members(i).Name(c) Then
                    varPrefix = varPrefix.Substring(0, c)
                    Exit For
                End If
            Next
            If varPrefix.Length = 0 Then Exit For
        Next


        b.BeginBlock("public enum {0}", enumName)
        For Each m In members
            Dim var = CSharp.ApplyStyle(m.Name.Substring(varPrefix.Length))
            b.AppendSummary(m.Comments)
            b.Append(var)
            If m.Value IsNot Nothing Then
                b.Append(" = unchecked((int){0})", m.Value)
            End If
            b.AppendLine(",")
        Next
        b.TrimRight()
        b.CutRight(1)
        b.AppendLine()
        b.EndBlock()
    End Sub

    Public Overrides ReadOnly Property IsCefEnumType As Boolean
        Get
            Return True
        End Get
    End Property

    Public Overrides ReadOnly Property AsCefEnumType As CefEnumType
        Get
            Return Me
        End Get
    End Property

End Class

File: CfxGenerator/ApiTypes/CefEnumType.cs
// Copyright (c) 2014-2017 Wolfgang Borgsmüller
// All rights reserved.
// 
// This software may be modified and distributed under the terms
// of the BSD license. See the License.txt file for details.

using System.Diagnostics;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

public class CefEnumType : ApiType {

    public class EnumMember {
        public readonly string Name;
        public readonly string Value;

        public readonly CommentNode Comments;

        public EnumMember(string name, string value, CommentNode Comments) {
            this.Name = name;
            if(!string.IsNullOrEmpty(value))
                this.Value = value;
            this.Comments = Comments;
        }

        public override string ToString() {
            if(Value != null) {
                return Name + " = " + Value;
            } else {
                return Name;
            }
        }
    }

    private readonly EnumMember[] members;

    private readonly CommentNode comments;

    public CefEnumType(Parser.EnumNode enumNode)
        : base(enumNode.Name) {
        members = new EnumMember[enumNode.Members.Count];
        for(var i = 0; i <= enumNode.Members.Count - 1; i++) {
            members[i] = new EnumMember(enumNode.Members[i].Name, enumNode.Members[i].Value, enumNode.Members[i].Comments);
        }
        comments = enumNode.Comments;
    }

    protected CefEnumType(string name)
        : base(name) {
    }

    public override string OriginalSymbol {
        get { return Name + "_t"; }
    }

    public override string PInvokeSymbol {
        get { return "int"; }
    }

    public override string PublicSymbol {
        get { return CSharp.ApplyStyle(CfxName); }
    }

    public override string RemoteCallSymbol {
        get { return "int"; }
    }

    public override string RemoteSymbol {
        get { return PublicSymbol; }
    }

    public string CfxName {
        get { return "cfx_" + Name.Substring(4); }
    }

    public override string PublicCallArgument(string var) {
        return string.Format("(int){0}", var);
    }

    public override string PublicWrapExpression(string var) {
        return string.Format("({0}){1}", PublicSymbol, var);
    }

    public override string PublicUnwrapExpression(string var) {
        return "(int)" + var;
    }

    public override string RemoteUnwrapExpression(string var) {
        return "(int)" + var;
    }

    public override string RemoteWrapExpression(string var) {
        return string.Format("({0}){1}", RemoteSymbol, var);
    }

    private static string[] additionalFlags = { "CfxSslContentStatus", "CfxV8PropertyAttribute", "CfxDragOperationsMask", "CfxFileDialogMode", "CfxJsonWriterOptions", "CfxTransitionType", "CfxV8AccessControl" };

    public override void EmitRemotePreCallStatements(CodeBuilder b, string var) {
        b.AppendLine("call.{0} = (int){0};", CSharp.Escape(var));
    }

    private static string GetEnumMemberValue(string originalValue) {
        switch(originalValue) {

            case "ERR_CERT_COMMON_NAME_INVALID":
                return "CertCommonNameInvalid";
            case "ERR_CERT_VALIDITY_TOO_LONG":
                return "CertValidityTooLong";
            case "UINT_MAX":
                return "unchecked((int)UInt32.MaxValue)";
            case "REFERRER_POLICY_CLEAR_REFERRER_ON_TRANSITION_FROM_SECURE_TO_INSECURE":
                return "ClearReferrerOnTransitionFromSecureToInsecure";
            default:
                return string.Format("unchecked((int){0})", originalValue);
        }
    }

    public void EmitEnum(CodeBuilder b) {
        var enumName = CSharp.ApplyStyle(CfxName);

        b.AppendSummaryAndRemarks(comments);

        if(Name.Contains("_flags") || additionalFlags.Contains(enumName)) {
            b.AppendLine("[Flags()]");
        }

        var prefixBuilder = new StringBuilder();
        var allEqual = true;

        do {
            char c = members[0].Name[prefixBuilder.Length];
            for(var i = 1; i <= members.Length - 1; i++) {
                if(c != members[i].Name[prefixBuilder.Length]) {
                    allEqual = false;
                    break;
                }
            }
            if(allEqual)
                prefixBuilder.Append(c);

        } while(allEqual);

        while(prefixBuilder.Length > 0 && prefixBuilder[prefixBuilder.Length - 1] != '_')
            --prefixBuilder.Length;

        b.BeginBlock("public enum {0}", enumName);
        foreach(var m in members) {
            var var = CSharp.ApplyStyle(m.Name.Substring(prefixBuilder.Length));

            if(char.IsDigit(var[0])) {
                switch(enumName) {
                    case "CfxScaleFactor":
                        var = "ScaleFactor" + var;
                        break;
                    default:
                        Debug.Assert(false);
                        break;
                }
            }

            b.AppendSummary(m.Comments);
            b.Append(var);
            if(m.Value != null) {
                b.Append(" = {0}", GetEnumMemberValue(m.Value));
            }
            b.AppendLine(",");
        }
        b.TrimRight();
        b.CutRight(1);
        b.AppendLine();
        b.EndBlock();
    }

    public override bool IsCefEnumType {
        get { return true; }
    }

    public override CefEnumType AsCefEnumType {
        get { return this; }
    }
}